### PR TITLE
GH-32947: [Go] Implement Concatenate for REE Array

### LINKED ISSUE #32947: [Go] Concatenate Implementation for REE Array
based on ARROW-17709

**Reporter**: [Matthew Topol](https://issues.apache.org/jira/browse/ARROW-17710) / @zeroshade
#### PRs and other links:
- [GitHub Pull Request #14126](https://github.com/apache/arrow/pull/14126)

<sub>**Note**: *This issue was originally created as [ARROW-17710](https://issues.apache.org/jira/browse/ARROW-17710). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by zeroshade**
* Closes: #32947

### CODE REVIEW

**Comment by zeroshade** [comment]
CC @zagto

File: go/arrow/array/concat_test.go [hunk 1]
```diff
@@ -568,3 +569,59 @@ func TestConcatDictionaryNullSlots(t *testing.T) {
 
 	assert.Truef(t, array.Equal(actual, expected), "got: %s, expected: %s", actual, expected)
 }
+
+func TestConcatRunLengthEncoded(t *testing.T) {
```
Comment (lidavidm) [inline go/arrow/array/concat_test.go:573]: I suppose:

- Can we test all three index types?
- Can we test that we error properly on overflow?
- Can we test that we can concatenate right up to the edge of overflow?

File: go/arrow/array/concat.go [hunk 2]
```diff
@@ -517,6 +523,50 @@ func concat(data []arrow.ArrayData, mem memory.Allocator) (arrow.ArrayData, erro
...
+			off := encoded.FindPhysicalOffset(d)
+
+			runs[i] = NewSliceData(d.Children()[0], int64(off), int64(off+plen))
+			defer runs[i].Release()
+			values[i] = NewSliceData(d.Children()[1], int64(off), int64(off+plen))
+			defer values[i].Release()
+
+			physicalLength, overflow = addOvf(physicalLength, int(plen))
+			if overflow {
+				return nil, fmt.Errorf("%w: run length encoded array length must fit into a 32-bit signed integer",
```
Comment (lidavidm) [inline go/arrow/array/concat.go:543]: Hmm, shouldn't this depend on the offset type?

Comment (lidavidm) [reply go/arrow/array/concat.go:543]: I suppose there are separate checks below, and since lengths are limited to 32-bit...

**Review by lidavidm — COMMENTED** [review]
nit: PR/Issue should be "REE" array right?

File: go/arrow/array/concat_test.go (hunk 1 above)
Comment (zeroshade) [reply go/arrow/array/concat_test.go:573]: added

File: go/arrow/array/concat.go (hunk 2 above)
Comment (lidavidm) [reply go/arrow/array/concat.go]: Fair enough. I suppose you could short circuit here earlier for int16, but as long as it properly fails either way.

**Review by lidavidm — APPROVED** [review]

**Comment by zeroshade** [comment]
I'll merge this at EOD if no one else has comments

**Pull request merged**